lib/tokens.js

```diff
@@ -0,0 +1,687 @@
+      // scope-resolved once
+      throw new BedrockError(
+        'Token already used.',
+        'NotAllowedError', {
+          public: true,
+          httpStatusCode: 400
+        });
+    }
+
+    // token is not yet resolved, attempt to resolve it for `requester`...
+
+    // create a pairwise token for the requester if one does not exist
+    const {pairwiseToken: {value: pairwiseToken}} = await _upsertPairwiseToken(
+      {internalId, requester});
+
+    // update requester's resolution info for the token batch
+    let bs;
+    if(entry) {
+      bs = new Bitstring({
+        buffer: await Bitstring.decodeBits({encoded: entry.list})
+      });
+    } else {
+      bs = new Bitstring({length: 256});
+      entry = resolution[resolution.length] = {requester};
+    }
+    bs.set(index, true);
+    entry.list = await bs.encodeBits();
+
+    // mark token as resolved
+    resolvedList.set(index, true);
+
+    // update token batch
+    const query = {
+      'tokenBatch.id': base64url.encode(batchId),
+      // ensure that no change is applied if another process resolved a
+      // token concurrently
+      'tokenBatch.resolvedList': encoded
+    };
+    const collection = database.collections['tokenization-tokenBatch'];
+    const result = await collection.updateOne(query, {
+      $set: {
+        'meta.updated': Date.now(),
+        'tokenBatch.resolvedList': await resolvedList.encodeBits(),
+        [`tokenBatch.resolution.${encodedRequester}`]: entry
+      }
+    }, database.writeOptions);
+    if(result.result.n === 0) {
+      // another process resolved a token concurrently, try again
+      continue;
+    }
+
+    // return new pairwise token
+    return {pairwiseToken};
+  }
+}
+
+/**
+ * Resolves a token to the `internalId` to which it is linked. This
+ * `internalId` should never be shared outside of the system that uses this
+ * module. This function is useful for obtaining the `internalId` associated
+ * with a token such that registered, encrypted documents associated with the
+ * `internalId` can be retrieved.
+ *
+ * @param {object} options - Options to use.
+ * @param {string} options.token - The token to resolve.
+ *
+ * @returns {object} An object with the `internalId`.
+ */
+export async function resolveToInternalId({token}) {
+  // parse token
+  const {batchId} = await _parseToken({token});
+
+  // get batch document
+  const {tokenBatch} = await _getBatch({id: batchId});
+  const {internalId} = tokenBatch;
+
+  // special case resolve to internal ID with flag set
+  return {internalId};
+}
+
+async function _createToken({
+  hmac, tokenVersion, tokenBatch, index, attributes
+} = {}) {
+  // decode batch ID
+  const batchId = base64url.decode(tokenBatch.id);
+
+  // get version options
+  const {id: version, options: {batchSaltSize}} = tokenVersion;
+
+  // build data to encrypt/wrap: batchId|index|aadSize|aad|padding
+  // total = 192-bits, 24 bytes
+  const toWrap = new Uint8Array(UNWRAPPED_SIZE);
```

Comment:
- no length and boundary checking done in this function
- math not adding up
- comment says 24 bytes
- array created with 24 byte constant
- batchId length could be other than 16 bytes which would effect the array size
- below is batchId=16 + index=1 + asize=1 + attrs=[0 to 8] (?) + padding.  size could be 18 to 26.  if using 8 bytes of attrs, last two bytes would get dropped
- also could also be trying to create negative amount of padding bytes

Reply:
Thanks for the review. Yeah, this will be changing given some recent conversations. Some of these details such as the attribute size will be pushed to the application layer instead of needing to be down here. It will be up to the application to determine how to express the size of the attributes in its own token format -- and we'll get the size from the Uint8Array representing the attributes that is preparsed and passed in. That still doesn't fix the math entirely (just gives us back a byte), but we will also be allowing for dynamic sizes here. We just have to ensure we use 64-bit boundaries for AES-KW. Anyway, this section will change soon and incorporate some of your comments and other decisions.